File: example-scripts/test_security.py
#!/usr/bin/env python3
"""Unit tests for security.py (stdlib unittest, zero dependencies).

Run: python3 test_security.py   (or: python3 -m unittest test_security)
"""
import unittest

from security import _fuzz_factor_for, secure_sustainability_report


def entry(period, energy=10.0, carbon=100.0, **extra):
    e = {"reporting-period": period, "energy-consumption": energy, "carbon-footprint": carbon}
    e.update(extra)
    return e


class SecureSustainabilityReportTests(unittest.TestCase):
    def test_sorts_ascending_by_reporting_period(self):
        data = [entry("2026-03"), entry("2026-01"), entry("2026-02")]
        out = secure_sustainability_report(data)
        self.assertEqual([r["reporting-period"] for r in out], ["2026-01", "2026-02", "2026-03"])

    def test_drops_sub_daily_entries_before_capping(self):
        subdaily = entry("2026-01-01T12:00:00Z")
        daily = [entry(f"2026-{m:02d}-01") for m in range(1, 3)]
        out = secure_sustainability_report([subdaily] + daily)
        self.assertEqual(len(out), 2)
        self.assertTrue(all(len(r["reporting-period"]) <= 10 for r in out))

    def test_caps_at_366_keeping_most_recent(self):
        data = [entry(f"2020-01-{(i % 28) + 1:02d}") for i in range(400)]
        # Ensure genuinely distinct, sorted-friendly periods across a wider span
        data = [entry(f"{2020 + i // 365}-{((i % 365) // 30) + 1:02d}-{(i % 28) + 1:02d}") for i in range(400)]
        out = secure_sustainability_report(data)
        self.assertLessEqual(len(out), 366)

    def test_deterministic_across_calls(self):
        data = [entry("2026-01"), entry("2026-02")]
        first = secure_sustainability_report(data)
        second = secure_sustainability_report(data)
        self.assertEqual(first, second)

    def test_noise_bounded_within_one_percent(self):
        data = [entry("2026-01", energy=100.0, carbon=1000.0)]
        out = secure_sustainability_report(data)
        self.assertGreaterEqual(out[0]["energy-consumption"], 99.0)
        self.assertLessEqual(out[0]["energy-consumption"], 101.0)
        self.assertGreaterEqual(out[0]["carbon-footprint"], 990.0)
        self.assertLessEqual(out[0]["carbon-footprint"], 1010.0)

    def test_negative_sentinel_never_noised(self):
        data = [entry("2026-01", energy=-1, carbon=500.0)]
        out = secure_sustainability_report(data)
        self.assertEqual(out[0]["energy-consumption"], -1)

    def test_single_fuzz_factor_keeps_scopes_consistent(self):
        data = [entry("2026-01", energy=10, carbon=300, **{"scope-1": 100, "scope-2": 100, "scope-3": 100})]
        out = secure_sustainability_report(data)
        total = out[0]["scope-1"] + out[0]["scope-2"] + out[0]["scope-3"]
        self.assertAlmostEqual(total, out[0]["carbon-footprint"], delta=0.05)

    def test_fuzz_factor_is_pure_function_of_period(self):
        self.assertEqual(_fuzz_factor_for("2026-01"), _fuzz_factor_for("2026-01"))
        self.assertNotEqual(_fuzz_factor_for("2026-01"), _fuzz_factor_for("2026-02"))

    def test_empty_input(self):
        self.assertEqual(secure_sustainability_report([]), [])


if __name__ == "__main__":
    unittest.main()

File: example-scripts/test_security.js
#!/usr/bin/env node
/**
 * Tests for security.js (plain Node `assert`, zero dependencies).
 * Run: node test_security.js
 */
const assert = require("assert");
const { secureSustainabilityReport, fuzzFactorFor } = require("./security.js");

const entry = (period, energy = 10, carbon = 100, extra = {}) => ({
  "reporting-period": period,
  "energy-consumption": energy,
  "carbon-footprint": carbon,
  ...extra,
});

let passed = 0;
function test(name, fn) {
  try {
    fn();
    console.log(`ok - ${name}`);
    passed++;
  } catch (err) {
    console.error(`FAIL - ${name}`);
    console.error(err);
    process.exitCode = 1;
  }
}

test("sorts ascending by reporting-period", () => {
  const out = secureSustainabilityReport([entry("2026-03"), entry("2026-01"), entry("2026-02")]);
  assert.deepStrictEqual(out.map((r) => r["reporting-period"]), ["2026-01", "2026-02", "2026-03"]);
});

test("drops sub-daily entries before capping", () => {
  const subdaily = entry("2026-01-01T12:00:00Z");
  const daily = [1, 2].map((m) => entry(`2026-0${m}-01`));
  const out = secureSustainabilityReport([subdaily, ...daily]);
  assert.strictEqual(out.length, 2);
  assert.ok(out.every((r) => String(r["reporting-period"]).length <= 10));
});

test("caps at 366 keeping most recent", () => {
  const data = Array.from({ length: 400 }, (_, i) => {
    const year = 2020 + Math.floor(i / 365);
    const month = Math.floor((i % 365) / 30) + 1;
    const day = (i % 28) + 1;
    return entry(`${year}-${String(month).padStart(2, "0")}-${String(day).padStart(2, "0")}`);
  });
  const out = secureSustainabilityReport(data);
  assert.ok(out.length <= 366);
});

test("deterministic across calls", () => {
  const data = [entry("2026-01"), entry("2026-02")];
  assert.deepStrictEqual(secureSustainabilityReport(data), secureSustainabilityReport(data));
});

test("noise bounded within ~1%", () => {
  const out = secureSustainabilityReport([entry("2026-01", 100, 1000)]);
  assert.ok(out[0]["energy-consumption"] >= 99 && out[0]["energy-consumption"] <= 101);
  assert.ok(out[0]["carbon-footprint"] >= 990 && out[0]["carbon-footprint"] <= 1010);
});

test("negative scope is noised and sign preserved", () => {
  // -03: no "not reported" sentinel; scope values MAY legitimately be negative
  // (removals / net accounting) and are noised like any other value —
  // multiplication preserves the sign and the sums.
  const out = secureSustainabilityReport([
    entry("2026-01", 100, 300, { "scope-1": -50, "scope-2": 250, "scope-3": 100 }),
  ]);
  const fuzz = fuzzFactorFor("2026-01");
  assert.notStrictEqual(fuzz, 1); // guard: noise is observable for this period
  assert.strictEqual(out[0]["scope-1"], parseFloat((-50 * fuzz).toFixed(2)));
  assert.notStrictEqual(out[0]["scope-1"], -50); // noise was applied
  assert.ok(out[0]["scope-1"] < 0); // sign preserved
  const total = out[0]["scope-1"] + out[0]["scope-2"] + out[0]["scope-3"];
  assert.ok(Math.abs(total - out[0]["carbon-footprint"]) < 0.05);
});

test("single fuzz factor keeps scopes consistent", () => {
  const out = secureSustainabilityReport([
    entry("2026-01", 10, 300, { "scope-1": 100, "scope-2": 100, "scope-3": 100 }),
  ]);
  const total = out[0]["scope-1"] + out[0]["scope-2"] + out[0]["scope-3"];
  assert.ok(Math.abs(total - out[0]["carbon-footprint"]) < 0.05);
});

test("fuzzFactorFor is a pure function of period", () => {
  assert.strictEqual(fuzzFactorFor("2026-01"), fuzzFactorFor("2026-01"));
  assert.notStrictEqual(fuzzFactorFor("2026-01"), fuzzFactorFor("2026-02"));
});

test("empty input", () => {
  assert.deepStrictEqual(secureSustainabilityReport([]), []);
});

console.log(`\n${passed} passed`);
if (process.exitCode) process.exit(1);
